### PR TITLE
Fix base of BokehJS class hierarchy 

### PR DESCRIPTION
**Pull request by bryevdv**
This PR adds `models/model` and `models/component` and has public facing BokehJS models inherit from one or the other. 

This implements a small part of #3651

### CODE REVIEW

**Comment by bryevdv** [comment]
Weird this builds locally for me after a `gulp clean` will look into Travis failure when I get to the office.

File: bokehjs/src/coffee/api.coffee [hunk 1]
```diff
@@ -19,6 +19,9 @@ module.exports = (Bokeh) ->
   Bokeh.Random        = require("./common/random")
   Bokeh.SVGColors     = require("./common/svg_colors")
 
+  # models
+  Component = require('./models/component')
```
Comment (mattpap) [inline bokehjs/src/coffee/api.coffee]: This is unnecessary. `Component` is unlikely to be used by the user.

Comment (bryevdv) [inline bokehjs/src/coffee/api.coffee]: Agreed, I will remove it

**Comment by mattpap** [comment]
It fails for me locally as well.

**Comment by mattpap** [comment]
You should import from `../models/model` (lower case) in `source/data_source.coffee`.

**Comment by bryevdv** [comment]
Yah, trying to reproduce on clean checkout

**Comment by mattpap** [comment]
No need to do a clean checkout, just `rm -rf bokehjs/build`. I don't use `gulp clean`, but it looks like it doesn't clean enough.

**Comment by bryevdv** [comment]
Why was that every working for me? Grr... also do you know any way to get better error messages when there are errors like that?  and definitely seems like `gulp clean` does not work correctly

**Comment by bryevdv** [comment]
Oh, it looks like `gulp clean` only works at the top level:

    BUILD_DIR = if typeof argv.buildDir == "string" then argv.buildDir else "./build"

**Comment by mattpap** [comment]
> Why was that every working for me?

It depends on the exact workflow. Maybe you just did `gulp dev-build`, but used later used minified version of bokeh (invalid at this point). Or caching, or ...

**Comment by mattpap** [comment]
> Oh, it looks like gulp clean only works at the top level:

So, actually `gulp clean` works for me. As to the top-level thing, there is silent assumption that gulp is run from within `bokehjs/` directory.

**Comment by bryevdv** [comment]
I do `gulp build` from further in the tree all the time, I had assumed `gulp clean` worked too, since it "worked"

**Comment by bryevdv** [comment]
OK this PR seems pretty self-contained I think I'd rather merge this and new start new work on top of it. @mattpap is there anything else you would think should go in this PR before merging?

File: bokehjs/test/test_common/test_document.coffee [hunk 2]
```diff
@@ -455,7 +455,7 @@ describe "Document", ->
     # double-check different results if we do include_defaults
     parsed_with_defaults = JSON.parse(d.to_json_string(include_defaults=true))
     attrs = parsed_with_defaults['roots']['references'][0]['attributes']
-    expect('tags' of attrs).to.be.equal true
+    #expect('tags' of attrs).to.be.equal true
```
Comment (mattpap) [inline bokehjs/test/test_common/test_document.coffee]: Why? I presume will be fixed in a further PR?

**Comment by mattpap** [comment]
LGTM

**Comment by mattpap** [comment]
> I do gulp build from further in the tree all the time (...)

I didn't realize `gulp` allows to do this. I just tried this and `gulp clean` sill works for me from e.g. `bokehjs/src` or deeper.

**Comment by bryevdv** [comment]
OK well it was just a speculation looking at the code, maybe I just thought I had `gulp clean` but really didn't. Anyway, merging now.

**Pull request merged**